source/dem/ray_tracing.cc

```diff
@@ -0,0 +1,809 @@
+  std::vector<Point<dim>> total_intersection_points;
+  total_intersection_points.reserve(photon_handler.n_global_particles());
+
+  // Map :
+  // <photon_iterator, (distance, intersection point<dim>, iterator to remove)>
+  ankerl::unordered_dense::map<
+    types::particle_index,
+    std::tuple<double, Point<dim>, Particles::ParticleIterator<dim>>>
+    photon_intersection_points_map;
+
+  photon_intersection_points_map.reserve(
+    photon_handler.n_locally_owned_particles());
+
+  // Update cell neighbors
+  find_cell_neighbors<dim, true>(triangulation,
+                                 cells_local_neighbor_list,
+                                 cells_ghost_neighbor_list);
+
+  // Particle don't move, thus we sort them at the end of a load balance.
+  // Photons are being sorted every pseudo time step, thus no need to sort them.
+  particle_handler.sort_particles_into_subdomains_and_cells();
+
+  // Exchange ghost particles
+  particle_handler.exchange_ghost_particles(true);
+
+  while (photon_handler.n_global_particles() != 0)
+    {
+      simulation_control->increment_iteration();
+
+      pcout << "Remaining photon : " << photon_handler.n_global_particles()
+            << std::endl;
+
+      // Load balancing (if needed)
+      load_balance();
+
+      // Since last pseudo time step, there is a high chance that the photons
+      // have changed cell considering the size of their displacement relative
+      // to the cell size. Thus, we sort them.
+      photon_handler.sort_particles_into_subdomains_and_cells();
+
+      // Loop over each local cell in the triangulation. To do this, we loop
+      // over each cell neighbor list. Each local cell has a list and the first
+      // iterator in the main cell itself.
+      for (auto cell_neighbor_list_iterator = cells_local_neighbor_list.begin();
+           cell_neighbor_list_iterator != cells_local_neighbor_list.end();
+           ++cell_neighbor_list_iterator)
+        {
+          // The main cell
+          const auto main_cell = cell_neighbor_list_iterator->begin();
+
+          // Photons in main cell
+          typename Particles::ParticleHandler<dim>::particle_iterator_range
+            photons_in_main_cell = photon_handler.particles_in_cell(*main_cell);
+
+          // Loop on each photon in the main cell
+          for (auto current_photon = photons_in_main_cell.begin();
+               current_photon != photons_in_main_cell.end();
+               ++current_photon)
+            {
+              // Get the photon properties. We need the properties to find the
+              // insertion position of the current photon. This will be used to
+              // differentiate between intersection points when there's more
+              // than one.
+              auto photon_properties = current_photon->get_properties();
+
+              // Current photon insertion point.
+              const Point<dim> photon_insertion_point = [&]() {
```

Comment:
This we do in so many locations one day we will have to make an inlined/static function for that instead of copy pasting it!